add.go

```diff
@@ -210,26 +210,45 @@ func (b *Builder) user(mountPoint string, userspec string) (specs.User, error) {
 }
 
 // setOwner sets the uid and gid owners of a given path.
-// If path is a directory, recursively changes the owner.
-func setOwner(path string, user specs.User) error {
-	fi, err := os.Stat(path)
+func setOwner(src, dest string, user specs.User) error {
+	fid, err := os.Stat(dest)
 	if err != nil {
-		return errors.Wrapf(err, "error reading %q", path)
+		return errors.Wrapf(err, "error reading %q", dest)
 	}
-	if fi.IsDir() {
-		err2 := filepath.Walk(path, func(p string, info os.FileInfo, we error) error {
-			if err3 := os.Lchown(p, int(user.UID), int(user.GID)); err3 != nil {
-				return errors.Wrapf(err3, "error setting ownership of %q", p)
-			}
-			return nil
-		})
+	if !fid.IsDir() || src == "" {
+		if err := os.Lchown(dest, int(user.UID), int(user.GID)); err != nil {
+			return errors.Wrapf(err, "error setting ownership of %q", dest)
+		}
+		return nil
+	}
+	err = filepath.Walk(src, func(p string, info os.FileInfo, we error) error {
+		relPath, err2 := filepath.Rel(src, p)
 		if err2 != nil {
-			return errors.Wrapf(err2, "error walking dir %q to set ownership", path)
+			return errors.Wrapf(err2, "error getting relative path of %q to set ownership on destination", p)
+		}
+		if relPath != "." {
+			absPath := filepath.Join(dest, relPath)
+			if err2 := os.Lchown(absPath, int(user.UID), int(user.GID)); err != nil {
+				return errors.Wrapf(err2, "error setting ownership of %q", absPath)
+			}
 		}
 		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "error walking dir %q to set ownership", src)
 	}
-	if err := os.Lchown(path, int(user.UID), int(user.GID)); err != nil {
-		return errors.Wrapf(err, "error setting ownership of %q", path)
+	return nil
+}
+
+// ensureDir creates a directory if it doesn't exist, setting ownership and permissions as passed by user and perm.
+func ensureDir(path string, user specs.User, perm os.FileMode) error {
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		if err := os.MkdirAll(path, perm); err != nil {
+			return errors.Wrapf(err, "error ensuring directory %q exists", path)
+		}
+		if err := os.Chown(path, int(user.UID), int(user.GID)); err != nil {
+			return errors.Wrapf(err, "error ensuring directory %q exists", path)
```

Comment:
SHouldn't this be
"error setting ownership of %q"

Reply:
Thanks for reviewing!